lib/ui/painting/paint.cc

```diff
@@ -331,6 +331,97 @@ bool Paint::sync_to(DisplayListBuilder* builder,
   return true;
 }
 
+void Paint::toDlPaint(DlPaint& paint) const {
+  if (isNull()) {
+    return;
+  }
+  FML_DCHECK(paint == DlPaint());
+
+  tonic::DartByteData byte_data(paint_data_);
+  FML_CHECK(byte_data.length_in_bytes() == kDataByteCount);
+
+  const uint32_t* uint_data = static_cast<const uint32_t*>(byte_data.data());
+  const float* float_data = static_cast<const float*>(byte_data.data());
+
+  Dart_Handle values[kObjectCount];
+  if (!Dart_IsNull(paint_objects_)) {
+    FML_DCHECK(Dart_IsList(paint_objects_));
+    intptr_t length = 0;
+    Dart_ListLength(paint_objects_, &length);
+
+    FML_CHECK(length == kObjectCount);
+    if (Dart_IsError(
+            Dart_ListGetRange(paint_objects_, 0, kObjectCount, values))) {
+      return;
+    }
+
+    Dart_Handle shader = values[kShaderIndex];
+    if (!Dart_IsNull(shader)) {
+      if (Shader* decoded = tonic::DartConverter<Shader*>::FromDart(shader)) {
+        auto sampling =
+            ImageFilter::SamplingFromIndex(uint_data[kFilterQualityIndex]);
+        paint.setColorSource(decoded->shader(sampling).get());
+      }
+    }
+
+    Dart_Handle color_filter = values[kColorFilterIndex];
+    if (!Dart_IsNull(color_filter)) {
+      ColorFilter* decoded =
+          tonic::DartConverter<ColorFilter*>::FromDart(color_filter);
+      paint.setColorFilter(decoded->dl_filter());
```

Comment:
The version of this in the SkPaint method will assume that filter() returns a valid shared ptr (it calls ->skia_object() without any null check). The dl_filter() method will protect against nullptr which seems to be unnecessary based on how the SkPaint version is written, but it will also de-share the pointer - returning DlCF* - which then needs to be re-shared in the setColorFilter method. I think overall that "paint.setColorFilter(decoded->filter())" would be faster as it can transfer the shared ptr without de/re-sharing it and it avoids what seems like an unnecessary null check.